source/macros.tex

```diff
@@ -473,8 +478,27 @@
 
 \lstnewenvironment{codeblock}{\CodeBlockSetup}{}
 
-% Left-align listings titles
 \makeatletter
+
+\newsavebox\@signalbox
+
+% Redefine \begin{codeblock} to look for an immediately-preceding
+% \begin{example}.
```

Comment:
Instead of redefining it, can we rename the listing environment and define our `codeblock` environment in terms of the listings environment? Or is that not possible due to the way that listings monkeys with the input stream?

Reply:
I don't know what you mean. We already have `\lstnewenvironment{codeblock}{\CodeBlockSetup}{}`, above.
The problem is that I can't put the "detect marker" code into \CodeBlockSetup, because that's too late and I can't get at the preceding box anymore. In order to get control earlier, I'm redefining the codeblock environment to get an earlier grab.

Reply:
What I'm asking is whether we can instead use (say) `\lstnewenvironment{codeblockimpl}....` and then define a new `codeblock` environment in terms of `codeblockimpl`. Modifying an existing environment like this, rather than composing this environment out of the existing one, is a hack that we should avoid if we can. (Maybe we can't; so be it.)

Reply:
Tried this:
```
\newenvironment{codeblock}
{%
  \ifhmode%                             no last box in vertical mode
    \setbox\@signalbox=\lastbox%        grab and consume last box
    \copy\@signalbox%                   typeset the last box again
    \ifhbox\@signalbox%
      \ifdim\wd\@signalbox=0pt%         zero-width box is our marker
        \def\codeblockbeginpenalty{100}%
      \fi%
    \fi%
  \fi%
\begin{codeblockimpl}}
{\end{codeblockimpl}}
```

Result:

! Extra }, or forgotten \endgroup.
<recently read> \egroup 
                        
l.228 ...( ) < > % : ; . ? * + - / ^ & | ~ ! = , @
                                                  \textbackslash@ " '


So, the environment nesting appears to negatively impact @ recognition. No idea why.